src/mame/heathkit/intr_cntrl.cpp

```diff
@@ -168,14 +174,58 @@ void z37_intr_cntrl::device_start()
 	m_irq_raised = false;
 }
 
-void z37_intr_cntrl::device_reset()
-{
-	heath_intr_cntrl::device_reset();
-}
-
 void z37_intr_cntrl::block_interrupts(uint8_t data)
 {
 	m_intr_blocked = bool(data);
 
 	update_intr_line();
 }
+
+
+heath_intr_socket::heath_intr_socket(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, HEATH_INTR_SOCKET, tag, owner, clock),
+	device_single_card_slot_interface(mconfig, *this),
+	m_irq_line(*this),
+	m_cntrl(nullptr)
+{
+}
+
+heath_intr_socket::~heath_intr_socket()
+{
+}
+
+void heath_intr_socket::device_start()
+{
+	m_cntrl = get_device();
+}
+
+IRQ_CALLBACK_MEMBER(heath_intr_socket::irq_callback)
+{
+	return get_instruction();
+}
+
+uint8_t heath_intr_socket::get_instruction()
+{
+	// assume NO-OP
+	uint8_t instr = 0x00;
+
+	if (m_cntrl)
+	{
+		 instr = m_cntrl->get_instruction();
+	}
+
+	return instr;
+}
+
+device_heath_intr_interface *heath_intr_socket::get_device()
+{
+	device_heath_intr_interface *const connected = dynamic_cast<device_heath_intr_interface *>(get_card_device());
+	if (connected)
+	{
+		return connected;
+	}
+	else
+	{
+		return nullptr;
+	}
+}
```

Comment:
You shouldn’t need this – `device_single_card_slot_interface` already has the glue to downlcast the connected device in its implementation of `get_card_device`.